pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java

```diff
@@ -202,43 +203,28 @@ private String getMultiStageQueryResponse(String query, String queryOptions, Htt
     QueryEnvironment queryEnvironment = new QueryEnvironment(new TypeFactory(new TypeSystem()),
         CalciteSchemaBuilder.asRootSchema(new PinotCatalog(_pinotHelixResourceManager.getTableCache())), null, null);
     List<String> tableNames = queryEnvironment.getTableNamesForQuery(query);
-    String brokerTenant;
+    List<String> instanceIds;
     if (tableNames.size() != 0) {
       List<TableConfig> tableConfigList = getListTableConfigs(tableNames);
       if (tableConfigList == null || tableConfigList.size() == 0) {
         return QueryException.getException(QueryException.TABLE_DOES_NOT_EXIST_ERROR,
             new Exception("Unable to find table in cluster, table does not exist")).toString();
       }
 
-      // When routing a query, there should be at least one common broker tenant for the table. However, the server
-      // tenants can be completely disjoint. The leaf stages which access segments will be processed on the respective
-      // server tenants for each table. The intermediate stages can be processed in either or all of the server tenants
-      // belonging to the tables.
-      brokerTenant = getCommonBrokerTenant(tableConfigList);
-      if (brokerTenant == null) {
+      // find the unions of all the broker tenant tags of the queried tables.
+      Set<String> brokerTenantsUnion = getBrokerTenantTagsUnion(tableConfigList);
```

Comment:
this extracts the union of all tags used by the tables from the query only @tibrewalpratik17

Reply:
Sounds good!